README.md
# Bin-X

## Installation

### Requirements

1. Install [FragGeneScan](https://sourceforge.net/projects/fraggenescan).
   ```bash
   wget -O tools/FragGeneScan1.31.tar.gz https://sourceforge.net/projects/fraggenescan/files/FragGeneScan1.31.tar.gz
   tar -xzf tools/FragGeneScan1.31.tar.gz -C tools
   cd tools/FragGeneScan1.31 && make clean && make fgs && cd ../..
   ./tools/FragGeneScan1.31/run_FragGeneScan.pl
   ```
2. Install [HMMER](http://hmmer.org/).
   ```bash
   wget -O tools/hmmer.tar.gz http://eddylab.org/software/hmmer/hmmer.tar.gz
   tar -xzf tools/hmmer.tar.gz -C tools
   cd tools/hmmer-3.3.2 && ./configure && make && cd ../..
   ./tools/hmmer-3.3.2/src/hmmalign -h
   ```
3. Install [kmer-counter](https://github.com/alexpreynolds/kmer-counter).
   ```bash
   git clone https://github.com/alexpreynolds/kmer-counter tools/kmer-counter
   cd tools/kmer-counter/ && make && cd ../..
   ./tools/kmer-counter/kmer-counter --help
   ```
4. Get marker-gene hmm file.
   ```bash
   wget -O tools/marker.hmm -nc https://raw.githubusercontent.com/sufforest/SolidBin/4c9b9ea7b8d8a0df1b772669872b69006c490e67/auxiliary/marker.hmm
   head ./tools/marker.hmm
   ```````

### Bin-X Installation

1. Install using `setup.py`.
   ```bash
   python setup.py install
   ```
2. Run following command to run the tool.
   ```
   bin_x run --contig [CONTIG_FASTA] --coverages [COVERAGE_TSV] --out [OUT_DIR]
   ```

## Development

1. First install [Poetry](https://python-poetry.org/docs/).
   ```bash
   curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py | python -
   ```
2. Install the requirements of the project by running
    ```bash
    poetry install
    ```
3. Run the tool via:
    ```bash
    poetry run python -m bin_x.bin_x
    ```
4. (Optional) Build the documentation via:
    ```bash
   cd docs
    poetry run sphinx-autobuild . _build/html --port 8001
    ```

Additionally, linting and type-checking are configured to this project. You may install the git-hooks for the formatters
via,

 ```bash
poetry run pre-commit install
 ```

Type-checking can be done as:

 ```bash
poetry run mypy .
 ```


CHANGELOG.md
# Changelog

## 0.0.2 (2021-08-09)

- Integrate normalized coverage matrix
- Change CLI input/output formats slightly
- Only split seed contigs
    - decreases binning slightly but drastically improves performance

## 0.0.1 (2021-07-08)

- Binning Scripts Initial Release


bin_x/cli/features.py
from configparser import SectionProxy
from pathlib import Path

import click
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

from bin_x.cli.utils import handle_error, reduce_dimensions_to_2d
from bin_x.core.config import USER_CONFIG
from bin_x.core.features.coverage import parse_coverages
from bin_x.core.features.kmer_count import count_kmers
from bin_x.core.features.preprocess import (
    filter_short_contigs,
    get_contig_lengths,
    split_contigs,
)
from bin_x.core.features.scm_gene import identify_marker_genomes


def _visualize_initial_bins(df: pd.DataFrame, out_png: Path):
    # Get only the required columns
    df_kmer_counts = df.drop(["CONTIG_NAME", "PARENT_NAME", "CLUSTER"], axis=1)
    df_2d = reduce_dimensions_to_2d(df_kmer_counts)

    # Draw and save the un-clustered and clustered points
    plt.figure(figsize=(15, 10))
    sns.scatterplot(x="X", y="Y", data=df_2d[df.CLUSTER == -1], label="No Category", color="black", alpha=0.1)
    for i in range(df.CLUSTER.max() + 1):
        sns.scatterplot(x="X", y="Y", data=df_2d[df.CLUSTER == i], label=f"Cluster {i}")
    plt.savefig(out_png)
    plt.clf()


def create_dataset(
    contig_fasta: Path,
    coverage_file: Path,
    operating_dir: Path,
    kmer_k: int = 4,
    kmer_counter_tool: str = "kmer_counter",
    short_contig_threshold: int = 1000,
    coverage_thresh: float = 0.4,
    select_percentile: float = 0.95,
    seed_contig_split_len: int = 10000,
) -> Path:
    """
    Create a dataset using the given input and configuration.

    :param contig_fasta: Contig file to use for kmer counting.
    :param coverage_file: Coverage file containing abundance information.
    :param operating_dir: Directory to write temp files to.
    :param kmer_k: k value of the kmers to count.
    :param kmer_counter_tool: kmer counter tool to use. (kmer_counter/seq2vec)
    :param short_contig_threshold: Threshold to filter the short contigs.
    :param coverage_thresh: Threshold for a hit to be considered for the seed frequency distribution.
    :param select_percentile: Percentile to use for selecting the number of seeds.
                For example, 0.5 will take the median number of seeds.
    :param seed_contig_split_len: Length to split the seed contigs.
    :return: Path of merged dataset with initial bins marked.
    """

    filtered_fasta = operating_dir / "filtered.fasta"
    split_fasta = operating_dir / "split.fasta"
    output_dataset_csv = operating_dir / "dataset.csv"
    plot_out_png = operating_dir / "plot.png"
    kmers_operation_dir = operating_dir / "kmers"
    scm_operation_dir = operating_dir / "scm"
    operating_dir.mkdir(parents=True, exist_ok=True)
    kmers_operation_dir.mkdir(parents=True, exist_ok=True)
    scm_operation_dir.mkdir(parents=True, exist_ok=True)

    # 01. Calculate coverages
    click.secho("01. Calculating coverages...", bold=True)
    df_coverages = parse_coverages(coverage_file)

    # 02. Remove short contigs
    click.secho(f"02. Removing short contigs below {short_contig_threshold}bp...", bold=True)
    contig_lengths = get_contig_lengths(contig_fasta)
    removed_contigs = filter_short_contigs(contig_fasta, filtered_fasta, threshold=short_contig_threshold)
    click.secho(f"Removed {len(removed_contigs)} (of {len(contig_lengths)}) short contigs", fg="green", bold=True)

    # 03. Perform single-copy marker gene analysis
    click.secho("03. Performing single-copy marker gene analysis...", bold=True)
    seed_clusters = identify_marker_genomes(
        filtered_fasta,
        contig_lengths,
        scm_operation_dir,
        coverage_thresh=coverage_thresh,
        select_percentile=select_percentile,
    )
    click.secho(f"Found {len(seed_clusters)} seeds", fg="green", bold=True)

    # 04. Identify seed contigs and split them
    click.secho(f"04. Splitting all contigs to contain {seed_contig_split_len}bp...", bold=True)
    sub_contigs = split_contigs(filtered_fasta, split_fasta, seed_clusters, split_len=seed_contig_split_len)
    click.secho(f"Found {len(sub_contigs)} contigs after splitting", fg="green", bold=True)

    # 05. Calculate normalized kmer frequencies
    click.secho(f"05. Calculating normalized kmer frequencies using {kmer_counter_tool}...", bold=True)
    df_kmer_freq = count_kmers(split_fasta, kmers_operation_dir, k=kmer_k, tool=kmer_counter_tool)

    # 06. Create a dataset with the initial cluster information
    click.secho("06. Creating a dataset with the initial cluster information...", bold=True)
    indexed_seed_clusters = zip(seed_clusters, range(len(seed_clusters)))
    df_seed_clusters = pd.DataFrame.from_records(indexed_seed_clusters, columns=["PARENT_NAME", "CLUSTER"])
    df_sub_contig = pd.DataFrame.from_records(list(sub_contigs.items()), columns=["CONTIG_NAME", "PARENT_NAME"])
    df_initial_clusters = pd.merge(df_sub_contig, df_seed_clusters, how="outer")
    df_initial_clusters = df_initial_clusters.fillna(-1)
    df_initial_clusters["CLUSTER"] = df_initial_clusters["CLUSTER"].astype(int)

    # 07. Merge all the features
    click.secho("07. Merging all the features...", bold=True)
    df_merged = pd.merge(df_initial_clusters, df_kmer_freq)
    df_merged = pd.merge(df_merged, df_coverages, left_on="PARENT_NAME", right_on="CONTIG_NAME")
    df_merged = df_merged.rename(columns={"CONTIG_NAME_x": "CONTIG_NAME"})
    df_merged = df_merged.drop("CONTIG_NAME_y", axis=1)
    df_merged.to_csv(output_dataset_csv, index=False)
    click.secho(f"Generated csv with shape {df_merged.shape}", fg="green", bold=True)
    click.secho(f"Dumped features CSV at {output_dataset_csv}", fg="green", bold=True)

    # 08. Visualize
    click.secho("08. Drawing plots...", bold=True)
    _visualize_initial_bins(df_merged, plot_out_png)

    return output_dataset_csv


def run_create_dataset(contig_fasta: Path, coverage_file: Path, operating_dir: Path, parameters: SectionProxy) -> Path:
    """
    Create a dataset using the given input and configuration.

    :param contig_fasta: Contig file to use for kmer counting.
    :param coverage_file: Coverage file containing abundance information.
    :param operating_dir: Directory to write temp files to.
    :param parameters: Parameters INI section.
    :return: Path of merged dataset with initial bins marked.
    """

    return create_dataset(
        contig_fasta=contig_fasta,
        coverage_file=coverage_file,
        operating_dir=operating_dir,
        kmer_k=int(parameters["KmerK"]),
        kmer_counter_tool=parameters["KmerCounterTool"],
        short_contig_threshold=int(parameters["ContigLengthFilterBp"]),
        coverage_thresh=float(parameters["ScmCoverageThreshold"]),
        select_percentile=float(parameters["ScmSelectPercentile"]),
        seed_contig_split_len=int(parameters["SeedContigSplitLengthBp"]),
    )


@click.command()
@click.option("--config", prompt="Configuration file", help="The INI File to use for tool configuration.", type=Path)
@click.option("--contigs", prompt="Contig file", help="The contig file to perform the binning operation.", type=Path)
@click.option("--coverages", prompt="Coverage file", help="The tab-seperated file with abundance data.", type=Path)
@click.option("--out", prompt="Output Directory", help="The output directory for the tool.", type=Path)
def main(config: Path, contigs: Path, coverages: Path, out: Path):
    try:
        USER_CONFIG.read(config)
        parameters = USER_CONFIG["PARAMETERS"]
        run_create_dataset(contigs, coverages, out, parameters)
    except Exception as e:
        handle_error(e)


if __name__ == "__main__":
    main()
